pd/src/components/staking.rs

```diff
@@ -49,16 +58,31 @@ impl Staking {
             total_delegations = total_changes.delegations.len(),
             total_undelegations = total_changes.undelegations.len(),
         );
-        // TODO: now the delegations and undelegations need to be grouped by validator ?
+
+        // now the delegations and undelegations need to be grouped by validator
+        let delegations_by_validator: BTreeMap<IdentityKey, Vec<Delegate>> = total_changes
+            .delegations
+            .into_iter()
+            .group_by(|d| d.validator_identity.clone())
+            .into_iter()
+            .map(|(k, v)| (k, v.collect()))
+            .collect();
+        let undelegations_by_validator: BTreeMap<IdentityKey, Vec<Undelegate>> = total_changes
+            .undelegations
+            .into_iter()
+            .group_by(|u| u.validator_identity.clone())
+            .into_iter()
+            .map(|(k, v)| (k, v.collect()))
+            .collect();
```

Comment:
I'm not sure this code is correct; looking at the [`group_by`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.group_by) docs, it groups *consecutive* runs of the same key, but we need to collect *all* the (un)delegations with that key, and that's not going to be possible without allocating (which the `group_by`) doesn't do.  I think this is going to drop all of the (un)delegations except the last consecutive runs by identity key.

Instead I think it'd be better to iterate over the (un)delegations directly, and then manually push into the right place with the Entry API, like (sketch):
```rust
let mut delegations_by_validator = BTreeMap::<IdentityKey, Vec<Delegate>::new();
for d in total_changes.delegations {
    delegations_by_validator.entry(d.validator_identity.clone())
        .or_default()
        .push(d);
}
```